### ISSUE TITLE
Proposal: automatically growable stack

### ISSUE THREAD

**Issue by densh** [scala-native/scala-native#1396]
Scala Native currently uses C's call stack as-is. This means that similarly to JVM we have a fixed-size stack. This can be problematic for some applications that heavily rely on recursion as they can may require more than a few megabytes of stack. To counter it on the JVM, community uses work-arounds such as trampolining which inevitably cause a major performance degradation. 

It would be interesting to see if we can adopt Go-style automatically growable stack instead. The application could start with some small-ish initial stack size and then grow it on-demand as more stack is used. This would require us to instrument function prologue to check if there is enough stack for a function call followed by a call to runtime to auto-grow the stack otherwise. 

Before we can proceed with this idea we need to have a prototype implementation to see what's the performance impact first. 

Related: [LLVM segmented stacks](https://llvm.org/docs/SegmentedStacks.html).

**Comment by shelby3** [comment]
> This would require us to instrument function prologue to check if there is enough stack for a function call

Could a [page guard](https://en.wikipedia.org/wiki/Memory_protection#Paged_virtual_memory) exception (c.f. [also](https://news.ycombinator.com/item?id=16317820#16321380)) be used more efficiently instead?

**Comment by densh** [comment]
@shelby3 Indeed, that's another way to do it. We need to try both approaches and see which one works best.

**Comment by shelby3** [comment]
Here is [another resource](https://stackoverflow.com/questions/35123250/stack-allocation-for-c-green-threads) I stumbled upon. I don’t know if that will apply since you’re using LLVM.
